### PR TITLE
Feature/KITOSUDV-2747

### PR DESCRIPTION
**Pull request by a-naskret**
## KITOS Pull request template
The following procedure dictates the steps needed before a Pull request can be merged into master.

- [x] **Implement**: 
      *All requirements are implemented and unit tests are green*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*
      
- [x] **Request review**: 
      *Tag whomever you wish to review your code*
      
- [x] **Review completed**: 
      *Reviewer ticks this box when review comments have been submitted*
      
- [x] **Changes**: 
      *PR owner and reviewer agrees on which changes must be made and the changes are committed.*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*

### CODE REVIEW

File: Infrastructure.DataAccess/Mapping/ItContractMap.cs [hunk 1]
```diff
@@ -58,6 +58,11 @@ public ItContractMap()
                 .WithMany(t => t.Children)
                 .HasForeignKey(d => d.ParentId)
                 .WillCascadeOnDelete(false);
+            
+            HasOptional(t => t.CriticalityType)
+                .WithMany(t => t.References)
+                .HasForeignKey(d => d.CriticalityId)
+                .WillCascadeOnDelete(false);
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Mapping/ItContractMap.cs:65]: the other choice type references do not have the additional "cascade" option... do you know if "false" is default.. if not, remove it.

File: Infrastructure.DataAccess/Migrations/Configuration.cs [hunk 2]
```diff
@@ -90,6 +90,8 @@ protected override void Seed(KitosContext context)
 
                 AddOptions<PurchaseFormType, ItContract>(context.PurchaseFormTypes, globalAdmin, "SKI", "SKI 02.18", "SKI 02.19", "Udbud", "EU udbud", "Direkte tildeling", "Annoncering");
 
+                AddOptions<CriticalityType, ItContract>(context.CriticalityTypes, globalAdmin, "1", "2", "3", "4", "5", "6", "7");
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Migrations/Configuration.cs:93]: Just add "Kritikalitet 1" and "Kritikalitet 2"

File: Presentation.Web/App_Start/MappingConfig.cs [hunk 3]
```diff
@@ -355,6 +360,7 @@ public MappingProfile()
             CreateMap<ItContract, ItContractDTO>()
                   .ForMember(dest => dest.AssociatedSystemUsages, opt => opt.MapFrom(src => src.AssociatedSystemUsages.Select(x => x.ItSystemUsage)))
                   .ForMember(dest => dest.AgreementElements, opt => opt.MapFrom(src => src.AssociatedAgreementElementTypes.Select(x => x.AgreementElementType)))
+                  .ForMember(dest => dest.CriticalityName, opt => opt.MapFrom(src => src.CriticalityType.Name))
```
Comment (mrjsawdk) [inline Presentation.Web/App_Start/MappingConfig.cs:363]: I think you need to look at how the other choice types are mapped because CriticalityName is not enough so no need to spell it out

File: Presentation.Web/Controllers/API/V2/External/ItContracts/ItContractCriticalityTypeV2Controller.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        /// <param name="organizationUuid">organization context for the criticality type availability</param>
+        /// <returns>A uuid and name pair with boolean to mark if the criticality type is available in the organization</returns>
+        [HttpGet]
+        [Route("{purchaseTypeUuid}")]
+        [SwaggerResponse(HttpStatusCode.OK, Type = typeof(RegularOptionExtendedResponseDTO))]
+        [SwaggerResponse(HttpStatusCode.BadRequest)]
+        [SwaggerResponse(HttpStatusCode.Unauthorized)]
+        [SwaggerResponse(HttpStatusCode.Forbidden)]
+        [SwaggerResponse(HttpStatusCode.NotFound)]
+        public IHttpActionResult Get([NonEmptyGuid] Guid purchaseTypeUuid, [NonEmptyGuid] Guid organizationUuid)
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V2/External/ItContracts/ItContractCriticalityTypeV2Controller.cs:52]: naming of first argument is off.. you updated the doc though.. align with that

File: Presentation.Web/Models/API/V1/ItContractDTO.cs [hunk 5]
```diff
@@ -83,5 +83,8 @@ public string ObjectOwnerFullName
         public IEnumerable<NamedEntityDTO> DataProcessingRegistrations { get; set; }
 
         public Guid Uuid { get; set; }
+        public int? CriticalityId { get; set; }
+        public string CriticalityName { get; set; }
+        public virtual CriticalityType CriticalityType { get; set; }
```
Comment (mrjsawdk) [inline Presentation.Web/Models/API/V1/ItContractDTO.cs:88]: wow is that the domain object you have there in the DTO?
that's a no go.. see how the other option types are handled

hint: To get automapper to work with this you must name it:

`CriticalityTypeName` and `CriticalityTypeId`
Then it will find the property and extract the name and id from it.. 
again, just look at how the others are used

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 6]
```diff
@@ -260,7 +265,8 @@
                                         "Supplier($select=Name)," +
                                         "AssociatedSystemUsages($expand=ItSystemUsage($select=Id;$expand=ItSystem($select=Name,Disabled)))," +
                                         "DataProcessingRegistrations($select=IsAgreementConcluded)," +
-                                        "LastChangedByUser($select=Name,LastName)";
+                                        "LastChangedByUser($select=Name,LastName)," +
+                                        "CriticalityType($select=Name)";
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:269]: Name is not a key, only select the key.. look in other overviews how to manage this.,.. basically you want the full options set delivered from the options service.. create a lookup of that and use the map to generate "filter options", as well as resolving the "names" based on the uid

See in It-system-overview.ts or the DPR overview. They use choice types in the grid

File: Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts [hunk 7]
```diff
@@ -92,6 +96,7 @@
                 $scope.contractTemplates = contractTemplates;
                 $scope.purchaseForms = purchaseForms;
                 $scope.procurementStrategies = procurementStrategies;
+                $scope.criticalities = criticalities;
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:99]: look in DPR how choice types are mapped to view models and updates are handled with service calls

File: Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.view.html [hunk 8]
```diff
@@ -108,6 +108,20 @@ <h4>{{contract.name}}</h4>
                         </div>
 
                     </div>
+                    <div class="col-xs-12"  data-ng-hide="!hasViewAccess">
+                        <div class="form-group">
+                            <select2-option data-id="contract-criticality"
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.view.html:113]: use the binding model as in DPR where choice types are used

File: Presentation.Web/app/models/it-contract/it-contract.ts [hunk 9]
```diff
@@ -59,6 +59,8 @@
         Children: Array<IItContract>;
         /** Gets or sets the chosen agreement elements. */
         AgreementElements: Array<IAgreementElement>;
+        CriticalityId: number,
+        CriticalityType: ICriticalityType,
```
Comment (mrjsawdk) [inline Presentation.Web/app/models/it-contract/it-contract.ts:63]: cleanup after the DTO in the backend has been changed

File: Tests.Integration.Presentation.Web/Tools/EntityOptionHelper.cs [hunk 10]
```diff
@@ -33,6 +33,7 @@ public static class ResourceNames
             public const string NoticePeriodMonthTypes = "TerminationDeadlineTypes";
             public const string HandoverTrialTypes = "HandoverTrialTypes";
             public const string ContractRoles = "ItContractRoles";
+            public const string CriticalityTypes = "CriticalityTypes";
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/Tools/EntityOptionHelper.cs:36]: Where is the integration test of the controller?

**Review by mrjsawdk — CHANGES_REQUESTED** [review]
@a-naskret Some comments and then a final conclusion that this story is only 50% done

File: Presentation.Web/app/components/data-processing/data-processing-registration-overview.controller.ts [hunk 11]
```diff
@@ -107,19 +89,19 @@
                                     replaceRoleQuery(parameterMap.$filter, getRoleKey(role), role.id);
                             });
 
-                            parameterMap.$filter = replaceOptionQuery(parameterMap.$filter,
+                            parameterMap.$filter = Helpers.OdataQueryHelper.replaceOptionQuery(parameterMap.$filter,
```
Comment (a-naskret) [inline Presentation.Web/app/components/data-processing/data-processing-registration-overview.controller.ts:92]: I moved this method to a helper, because I wanted to reuse it in the ItContractOverview, but I dropped the idea later as it wouldn't work.

Should I revert this change?

File: Core.ApplicationServices/Model/Contracts/Write/ItContractGeneralDataModificationParameters.cs [hunk 12]
```diff
@@ -15,5 +15,7 @@ public class ItContractGeneralDataModificationParameters
         public OptionalValueChange<Maybe<bool>> EnforceValid { get; set; } = OptionalValueChange<Maybe<bool>>.None;
         public OptionalValueChange<Maybe<DateTime>> ValidFrom { get; set; } = OptionalValueChange<Maybe<DateTime>>.None;
         public OptionalValueChange<Maybe<DateTime>> ValidTo { get; set; } = OptionalValueChange<Maybe<DateTime>>.None;
+        public OptionalValueChange<string> CriticalityTypeId { get; set; } = OptionalValueChange<string>.None;
```
Comment (mrjsawdk) [inline Core.ApplicationServices/Model/Contracts/Write/ItContractGeneralDataModificationParameters.cs:18]: what is this?.. I understand the uuid but not the string

File: Infrastructure.DataAccess/Mapping/CriticalityTypeMap.cs [hunk 13]
```diff
@@ -0,0 +1,8 @@
+﻿using Core.DomainModel.ItContract;
+
+namespace Infrastructure.DataAccess.Mapping
+{
+    public class CriticalityTypeMap : OptionEntityMap<CriticalityType, ItContract>
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Mapping/CriticalityTypeMap.cs:5]: TODO MRJ: Check if the local** needs a map

File: Presentation.Web/Controllers/API/V2/External/ItContracts/Mapping/ItContractWriteModelMapper.cs [hunk 14]
```diff
@@ -304,7 +304,15 @@ public IEnumerable<ItContractHandoverTrialUpdate> MapHandOverTrials(IEnumerable<
 
                 EnforceValid = rule.MustUpdate(x => x.General.Validity.EnforcedValid)
                     ? (dto.Validity?.EnforcedValid ?? Maybe<bool>.None).AsChangedValue()
-                    : OptionalValueChange<Maybe<bool>>.None
+                    : OptionalValueChange<Maybe<bool>>.None,
+
+                CriticalityTypeId = rule.MustUpdate(x => x.General.CriticalityTypeId)
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V2/External/ItContracts/Mapping/ItContractWriteModelMapper.cs:309]: why is this here????... why two versions of criticality

File: Presentation.Web/Models/API/V2/Request/Contract/ContractGeneralDataWriteRequestDTO.cs [hunk 15]
```diff
@@ -37,5 +37,16 @@ public class ContractGeneralDataWriteRequestDTO
         /// Validity of the it-contract
         /// </summary>
         public ValidityWriteRequestDTO Validity { get; set; }
+        /// <summary>
+        /// User-assigned criticality type id
```
Comment (mrjsawdk) [inline Presentation.Web/Models/API/V2/Request/Contract/ContractGeneralDataWriteRequestDTO.cs:41]: again - why is this here`?

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 16]
```diff
@@ -174,6 +177,11 @@
             return filterUrl.replace(pattern, "AssociatedSystemUsages/any(c: $1c/ItSystemUsage/ItSystem/Name$2)");
         }
 
+        private fixCriticalityFilter(filterUrl, column) {
+            const pattern = new RegExp(`(\\w+\\()${column}(.*?\\))`, "i");
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:181]: dead, unused code?

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 17]
```diff
@@ -261,16 +269,25 @@
...
+                                    query = `/odata/Organizations(${self.user.currentOrganizationId})/ItContracts`;
                                 } else {
-                                    return `/odata/Organizations(${self.user
-                                        .currentOrganizationId})/OrganizationUnits(${orgUnitId})/ItContracts` +
-                                        urlParameters;
+                                    query = `/odata/Organizations(${self.user
+                                        .currentOrganizationId})/OrganizationUnits(${orgUnitId})/ItContracts`;
                                 }
+
+                                var criticalityId = self.$window.sessionStorage.getItem(self.criticalityTypeStorageKey);
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:283]: what?... why this odd approach. Kendo will generate the right filter query.

Comment (mrjsawdk) [reply Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:283]: and that you must replace using a regex

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 18]
```diff
@@ -261,16 +269,25 @@
...
+                                if (criticalityId === null) {
+                                    urlParameters += ",CriticalityType($select=Id)";
+                                } else {
+                                    urlParameters += `,CriticalityType($select=Id)&%24filter=CriticalityType/Id eq ${criticalityId}`;
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:287]: this is not how we do it.. look in the other controllers
expansion criteria work regardless of the filter..

also filter must be applied using the filter query and mutations to that. Remember that last time we went over the same problem

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 19]
```diff
@@ -938,6 +977,15 @@
 
             return concatRoles;
         }
+
+        private getCriticalityName(id: number): string {
+            const criticalitiesWithSameId = this.criticalityOptions.filter(x => x.Id === id);
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:982]: searching every time when you don't know how many times you have to do it = bad performance.. build a dictionary/object map, like we do with enums and use that for lookups

File: Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts [hunk 20]
```diff
@@ -307,6 +328,54 @@
...
+                    let existingChoice = null;
+                    if ($scope.contract.criticalityTypeId !== undefined && $scope.contract.criticalityTypeId !== null) {
+                        existingChoice = {
+                            id: $scope.contract.criticalityTypeId,
+                            name: $scope.contract.criticalityTypeName
+                        };
+
+                        if (!optionMap[$scope.contract.criticalityTypeId]) {
+                            optionMap[existingChoice.id] = {
+                                text: existingChoice.name,
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:356]: you are missing something here
image

File: Presentation.Web/app/models/it-contract/criticality-type.ts [hunk 21]
```diff
@@ -0,0 +1,8 @@
+﻿module Kitos.Models.ItContract {
+    export interface ICriticalityType extends IEntity {
+        Name: string;
+        IsActive: boolean;
+        Note: string;
+        References: Array<IItContract>;
```
Comment (mrjsawdk) [inline Presentation.Web/app/models/it-contract/criticality-type.ts:6]: do you need that in the ui dto?

File: Presentation.Web/app/models/it-contract/it-contract.ts (hunk 9 above)
Comment (mrjsawdk) [reply Presentation.Web/app/models/it-contract/it-contract.ts:66]: This one is the OData model for the overview so if you are using the properties then keep these,

File: Presentation.Web/app/models/it-contract/criticality-type.ts (hunk 21 above)
Comment (mrjsawdk) [reply Presentation.Web/app/models/it-contract/criticality-type.ts:6]: do you need this specific dto? Isn't there an OptionDTO?

File: Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs [hunk 22]
```diff
@@ -99,7 +99,7 @@ public static IEnumerable<object[]> GetUndefinedSectionsInput()
 
         public static IEnumerable<object[]> GetUndefinedGeneralDataPropertiesInput()
         {
-            return CreateGetUndefinedSectionsInput(8);
+            return CreateGetUndefinedSectionsInput(10);
```
Comment (mrjsawdk) [inline Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs:102]: again this additional type bumps the number from 8-10 in stead of 0

File: Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs [hunk 23]
```diff
@@ -227,11 +227,13 @@ public static IEnumerable<object[]> GetUndefinedTerminationDataPropertiesInput()
             bool noNotes,
             bool noEnforceValid,
             bool noValidFrom,
-            bool noValidTo)
+            bool noValidTo,
+            bool noCriticalityTypeId,
```
Comment (mrjsawdk) [inline Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs:231]: kill with 🔥

File: Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs [hunk 24]
```diff
@@ -257,11 +259,13 @@ public static IEnumerable<object[]> GetUndefinedTerminationDataPropertiesInput()
...
```
Comment (mrjsawdk) [inline Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs:263]: kill with 🔥

File: Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs [hunk 25]
```diff
@@ -1596,7 +1604,9 @@ private static void AssertPaymentCollection(IEnumerable<PaymentRequestDTO> expec
...
```
Comment (mrjsawdk) [inline Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs:1608]: kill with 🔥

File: Core.ApplicationServices/Model/Contracts/Write/ItContractGeneralDataModificationParameters.cs [hunk 26]
```diff
@@ -15,5 +15,7 @@ public class ItContractGeneralDataModificationParameters
...
+        public OptionalValueChange<Guid?> CriticalityTypeUuid { get; set; } = OptionalValueChange<Guid?>.None;
```
Comment (mrjsawdk) [inline Core.ApplicationServices/Model/Contracts/Write/ItContractGeneralDataModificationParameters.cs:19]: none of these two properties are used in the write service, so how can this be ready for review?

**Review by mrjsawdk — CHANGES_REQUESTED** [review]
@a-naskret Still not close to being done

**Like the last time:**
- [ ]  The criticality type cannot be changes via the API V2. You added two properties (there should only be one) but you are not using them for anything.
- [ ] The criticality type is missing from the response model. It's only in the write model. See how the others are represented and mapped
- [ ]  There is no extension to the integration test (for V2) using the new criticality type, and that should warn you, that you are not done.
- There is also no extension to the integration test for V1 - there should be
- [ ]  Filtering in kendo is handled using a new odd approach. Please lean on existing solutions
- [ ]  State in the view is flaky. Try opening a new contract, change criticality, go to another tab and then back again. Notice how the select box is empty. Look how the other options are handled like in DPR and how state is preserved correctly
- [ ]  Filter state does not update correctly in kendo: 
image
- [ ]  Filter state does not work if "blank" is selected, which really points back to the main issue with the way filtering is handled

- [ ]  "ordering" in kendo does not work... ordering should update the order query to use the name

I will pull this one back into progress

File: Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts [hunk 27]
```diff
@@ -307,6 +328,54 @@
...
+                        }
+                    }
+
+                    const options = criticalityOptions.map(option => optionMap[option.Id]);
+
+                    $scope.criticality = {
+                        selectedElement: existingChoice && optionMap[existingChoice.id],
+                        select2Config: select2LoadingService.select2LocalDataNoSearch(() => options, true),
+                        elementSelected: (newElement) => {
+                            if (!newElement)
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:370]: this is why clearing the selection does not work

Comment (mrjsawdk) [reply Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:370]: See how it's done in DPR 
image

File: Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts [hunk 28]
```diff
@@ -307,6 +328,52 @@
                     }
                 }
 
+                function bindCriticalities() {
+                    
+                    const optionMap = criticalityOptions.reduce((acc, next, _) => {
+                        acc[next.Id] = {
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:334]: I would bet we could refactor this block (lin 334-344) into a helper since it works on the generic optionType `Kitos.Models.IOptionEntity`

Then we can save 10 lines for each time we need to reduce a list of option types into a lookup dictionary containing those select 2 options

File: Presentation.Web/app/helpers/kendo-overview-helper.ts.ts [hunk 29]
```diff
@@ -0,0 +1,30 @@
+﻿module Kitos.Helpers {
+
+    export class KendoOverviewHelper {
+
+        /**
+          *  creates a select dropdown for given dataSource
+          * @param args - should contain kenoGrid elements
+          * @param dataSource - dataSource of the dropdown, should contain 'id' and 'text' properties
```
Comment (mrjsawdk) [inline Presentation.Web/app/helpers/kendo-overview-helper.ts.ts:8]: in stead of writing it in text, create an interface type and replace `any`with that in the argument list

File: Presentation.Web/app/helpers/kendo-overview-helper.ts.ts [hunk 30]
```diff
@@ -0,0 +1,30 @@
+﻿module Kitos.Helpers {
+
+    export class KendoOverviewHelper {
+
+        /**
+          *  creates a select dropdown for given dataSource
+          * @param args - should contain kenoGrid elements
```
Comment (mrjsawdk) [inline Presentation.Web/app/helpers/kendo-overview-helper.ts.ts:7]: also an any... please provide a real type of what you expect

File: Presentation.Web/app/helpers/kendo-overview-helper.ts.ts [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+          * @param dataSource - dataSource of the dropdown, should contain 'id' and 'text' properties
+          * @param insertEmptyOption - if true and if dataSource doesn't contain an item with an empty name, method will add an empty option to select
+          */
+        public static createSelectDropdownTemplate(args: any, dataSource: any, insertEmptyOption: boolean) {
+            if (insertEmptyOption && dataSource.filter(x => x.text === "").length === 0) {
```
Comment (mrjsawdk) [inline Presentation.Web/app/helpers/kendo-overview-helper.ts.ts:12]: is text === "" a key?... why not add an additional property to the selection in stead called "emptyOption"?

File: Presentation.Web/app/helpers/kendo-overview-helper.ts.ts [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+                dataSource.push({ id: 0, text: "" });
```
Comment (mrjsawdk) [inline Presentation.Web/app/helpers/kendo-overview-helper.ts.ts:13]: this adds is to the end of the list which is strange. Place it as the head element in stead

File: Presentation.Web/app/helpers/kendo-overview-helper.ts.ts [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+
+    export class KendoOverviewHelper {
+
+        /**
+          *  creates a select dropdown for given dataSource
+          * @param args - should contain kenoGrid elements
+          * @param dataSource - dataSource of the dropdown, should contain 'id' and 'text' properties
+          * @param insertEmptyOption - if true and if dataSource doesn't contain an item with an empty name, method will add an empty option to select
+          */
+        public static createSelectDropdownTemplate(args: any, dataSource: any, insertEmptyOption: boolean) {
```
Comment (mrjsawdk) [inline Presentation.Web/app/helpers/kendo-overview-helper.ts.ts:11]: please also add types for the response

File: Presentation.Web/app/models/ViewModel/it-contract/contract-criticality-view-model.ts [hunk 34]
```diff
@@ -0,0 +1,30 @@
...
+    export class CriticalityOptions {
+
+        private getValueToTextMap() {
+            var result = {};
+            this.options.forEach(value => result[value.id] = value.text);
+            return result;
+        }
+
+        //Cache texts for quick lookup texts
+        private readonly valueToTextMap;
```
Comment (mrjsawdk) [inline Presentation.Web/app/models/ViewModel/it-contract/contract-criticality-view-model.ts:13]: please define a type for it. I would suggest:
`Record<number,string>`

File: Presentation.Web/app/models/generic/option-dto.ts [hunk 35]
```diff
@@ -0,0 +1,9 @@
+﻿module Kitos.Models.OData.Generic {
+
+    export interface IOptionDTO<TReference extends IEntity> extends IEntity {
```
Comment (mrjsawdk) [inline Presentation.Web/app/models/generic/option-dto.ts:3]: perhaps let's put it in an odata sub folder (the module namespace is fine).. just to make it easier to find

File: Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts [hunk 36]
```diff
@@ -976,6 +1021,12 @@
                         ],
                         uiState: [
                             "uiCustomizationStateService", (uiCustomizationStateService: Kitos.Services.UICustomization.IUICustomizationStateService) => uiCustomizationStateService.getCurrentState(Kitos.Models.UICustomization.CustomizableKitosModule.ItContract)
+                        ],
+                        criticalityOptions: [
+                            "localOptionServiceFactory",
+                            (localOptionServiceFactory: Services.LocalOptions.ILocalOptionServiceFactory) =>
+                            localOptionServiceFactory.create(Services.LocalOptions.LocalOptionType.CriticalityTypes)
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:1028]: this endpoint only returns the currently available options which is fine for the selection box options, but since we use it for lookup it will fail (result in blank) on options which are registered and then expired by the local admin. (in the controller it is managed with the "existingOption". See in DPR's overview for "Overførselsgrundlag" on how to deal with this.

Comment (mrjsawdk) [reply Presentation.Web/app/components/it-contract/it-contract-overview.controller.ts:1028]: Hint: It uses a custom endpoint which exports all needed types including their expiration status (uses generic services for it so it's easy to build)

File: Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts [hunk 37]
```diff
@@ -307,6 +328,52 @@
...
+                        }
+                    }
+
+                    const options = criticalityOptions.map(option => optionMap[option.Id]);
+
+                    $scope.criticality = {
+                        selectedElement: existingChoice && optionMap[existingChoice.id],
+                        select2Config: select2LoadingService.select2LocalDataNoSearch(() => options, true),
+                        elementSelected: (newElement) => {
+                            var payload = { criticalityTypeId: newElement ? newElement.id : null };
```
Comment (mrjsawdk) [inline Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:370]: still does not work when the option is cleared (using the `x`)

Comment (mrjsawdk) [reply Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:370]: image

Comment (mrjsawdk) [reply Presentation.Web/app/components/it-contract/tabs/it-contract-tab-main.controller.ts:370]: Hint: Works fine in DPR - Overførselsgrundlag

File: Tests.Integration.Presentation.Web/Contract/ItContractTest.cs [hunk 38]
```diff
@@ -94,6 +94,44 @@ public async Task Cannot_Add_PaymentMilestone(OrganizationRole role)
...
+
+        [Theory]
+        [InlineData(OrganizationRole.User)]
+        public async Task Cannot_Add_CriticalityType(OrganizationRole role)
+        {
+            //Arrange
+            var login = await HttpApi.GetCookieAsync(role);
+            var contract = await ItContractHelper.CreateContract(A<string>(), OrganizationId);
+            var criticalityName = A<string>();
+            var criticality = await EntityOptionHelper.CreateOptionTypeAsync(EntityOptionHelper.ResourceNames.CriticalityTypes, criticalityName, OrganizationId);
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/Contract/ItContractTest.cs:126]: why create a new option type in this test... is it not enough just to query the existing and pick a random one?

File: Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs [hunk 39]
```diff
@@ -686,17 +686,17 @@ public async Task Can_PATCH_With_Procurement()
...
-        [InlineData(false, true, true)]
-        [InlineData(true, true, true)]
-        [InlineData(false, false, false)]
-        public async Task Can_POST_With_GeneralData(bool withContractType, bool withContractTemplate, bool withAgreementElements)
+        [InlineData(true, true, true, true)]
+        [InlineData(true, true, true, false)]
+        [InlineData(true, true, false, true)]
+        [InlineData(true, false, true, true)]
+        [InlineData(false, true, true, true)]
+        [InlineData(false, false, false, false)]
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs:694]: Remember to also cover the PUT and PATCH scenarios (see the rest of this file)

File: Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs [hunk 40]
```diff
@@ -227,11 +227,12 @@ public static IEnumerable<object[]> GetUndefinedTerminationDataPropertiesInput()
...
             bool noEnforceValid,
             bool noValidFrom,
-            bool noValidTo)
+            bool noValidTo,
+            bool noCriticalityTypeUuid)
         {
             //Arrange
             var input = new CreateNewContractRequestDTO();
-            ConfigureGeneralDataInputContext(noContractId, noContractTypeUuid, noContractTemplateUuid, noAgreementElementUuids, noNotes, noEnforceValid, noValidFrom, noValidTo);
+            ConfigureGeneralDataInputContext(noContractId, noContractTypeUuid, noContractTemplateUuid, noAgreementElementUuids, noNotes, noEnforceValid, noValidFrom, noValidTo, noCriticalityTypeUuid);
```
Comment (mrjsawdk) [inline Tests.Unit.Presentation.Web/Models/V2/ItContractWriteModelMapperTest.cs:235]: you are changing the input in all of these tests but there are no changes to the assert blocks.

Please fix this

**Review by mrjsawdk — CHANGES_REQUESTED** [review]
@a-naskret Still some work to do here but closer to the goal.

Let's get this in good shape

File: Tests.Integration.Presentation.Web/Tools/EntityOptionHelper.cs [hunk 41]
```diff
@@ -105,5 +107,15 @@ public static async Task<HttpResponseMessage> SendChangeOptionIsObligatoryAsync(
             Assert.Equal(HttpStatusCode.NoContent, response.StatusCode);
             return response;
         }
+
+        public static async Task<List<OptionDTO>> GetOptionsAsync(string resource, Cookie optionalLogin = null)
```
Comment (a-naskret) [inline Tests.Integration.Presentation.Web/Tools/EntityOptionHelper.cs:111]: I had to add a new method in order to get the options, because the V2 Helper method required an organizationUuid, so I thought it's unnecessary to create an organization just to get the options

File: Presentation.Web/app/helpers/option-entity-helper.ts [hunk 42]
```diff
@@ -0,0 +1,20 @@
+﻿module Kitos.Helpers {
+
+    export class OptionEntityHelper {
+
+        static createDictionaryFromOptionList(criticalities: Kitos.Models.IOptionEntity[]) {
```
Comment (mrjsawdk) [inline Presentation.Web/app/helpers/option-entity-helper.ts:5]: naming og the argument should be `options` :-)

**Review by mrjsawdk — COMMENTED** [review]
@a-naskret Looks super nice.
Once you finish the last part about the options endpoint, let's do a quick final check and then this one is done. 

Great work 👍

File: Presentation.Web/Controllers/API/V1/ItContractController.cs [hunk 43]
```diff
@@ -388,5 +409,15 @@ public override HttpResponseMessage Post(int organizationId, ItContractDTO dto)
...
         }
+
+        private static IEnumerable<OptionWithDescriptionAndExpirationDTO> ToDTOs<T>(IEnumerable<(OptionDescriptor<T> option, bool available)> options, int organizationId) where T : OptionEntity<ItContract>
+        {
+            return options.Select(ToDTO);
+        }
+
+        private static OptionWithDescriptionAndExpirationDTO ToDTO<T>((OptionDescriptor<T> option, bool available) optionObject) where T : OptionEntity<ItContract>
+        {
+            return new OptionWithDescriptionAndExpirationDTO(optionObject.option.Option.Id, optionObject.option.Option.Name, optionObject.available, optionObject.option.Description);
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V1/ItContractController.cs:420]: Notice.. you are putting available into an "expired" argument. It must be `!available`

File: Presentation.Web/app/models/ViewModel/it-contract/contract-criticality-view-model.ts [hunk 44]
```diff
@@ -0,0 +1,49 @@
...
+            if (takeUnavailableText) {
+                return value.available
+                    ? value.text
+                    : value.unavailableText;
+            }
+
+            return value.text;
+        }
+
+        options: Select2OptionViewModel<any>[];
```
Comment (mrjsawdk) [inline Presentation.Web/app/models/ViewModel/it-contract/contract-criticality-view-model.ts:37]: you can actually set it to the dto type in stead of any

**Review by mrjsawdk — APPROVED** [review]
